### DISCUSSION TITLE
Code Block Formatting/Fixers?

### DISCUSSION THREAD

**Discussion by acederberg** [quarto-dev/quarto-cli#11063] [Q&A]
### Description

Hi everybody! Thank you again for this wonderful project.

## Use Case

I was wondering if there is any known way to clean up poorly formatted documents. For instance, using injected formatters with [conform.nvim](https://github.com/stevearc/conform.nvim) I can get ``python``, ``yaml``, and other languages fixed in my notebook through ``nvim``. I would like to run this on some files in bulk in github actions - however this might be quite tricky or impossible with ``nvim``.

I would like to note that I do not think the ``quarto`` command itself should do this, but am asking if there is a solution that already exists and if not, how to proceed possibly implementing a solution.

As for my thoughts, I think it might be possible using ``pandoc`` filters and updating the files in place. 

Apparently, [prettier](https://prettier.io/blog/2017/11/07/1.8.0.html) is capable of fixing code blocks in markdown however only ``js`` based plugins can be used for this.

## What Would it Do?

This functionality could turn the following poorly formatted ``quarto`` document

~~~quarto
---
title: Poorly formatted front matter
categories:
- poorly
- formated
---

blah blah blah

```{.python}
print( "poorly" , "formated"
)
```
~~~

into 

~~~quarto
---
title: Poorly formatted front matter
categories: [poorly, formated]
---

blah blah blah

```{python}
print("poorly", "formated")
```
~~~

when the user specifies to use ``black`` and ``yamlfix``.

**Comment by mcanouil** [comment]
> I would like to note that I do not think the quarto command itself should do this, but am asking if there is a solution that already exists and if not, how to proceed possibly implementing a solution.

That's very unlikely as Quarto supports any possible programming languages either via `knitr` or Jupyter kernels.
Quarto cannot possibly embed languages and formatter.
As a reminder, Quarto is language agnostic. Using R, Python, Julia, etc. is a user action so is the formatting.

This being said, you can build your own system as some users already did, see <https://github.com/shafayetShafee/black-formatter> for Black (Python).

Any formatter and their settings is opinionated so it's best you do what you think is best in that regard.

**Comment by acederberg** [comment]
Thanks!

> This being said, you can build your own system as some users already did, see https://github.com/shafayetShafee/black-formatter for Black (Python).

Something like that would be cool. I think I might implement a ``pandoc`` filter that:

1. Can be configured to use certain tools on certain languages and do nothing on others, e.g. only use ``yamlfix`` on ``yaml`` codeblocks. 
2. Configuration of the particular formatter should be left to the user and not configured in the filter configuration itself. E.g. in ``pyproject.toml`` for ``yamlfix`` and ``black``.
3. Outputs the reformatted code into the respective ``qmd`` file instead of outputting to ``html`` so that source files are fixed and not just the output.

One final question. To make this work, is it possible to load a ``qmd`` file and output it to a ``qmd`` file with ``pandoc``? I tried 

```sh
quarto pandoc -o ./tables.qmd ./tables.qmd
```

and got 

```stderr
[WARNING] Could not deduce format from file extension .qmd
  Defaulting to markdown
[WARNING] Could not deduce format from file extension .qmd
  Defaulting to html
```

**Reply by mcanouil** [reply]
qmd is "markdown with code cells (computation)".
If Pandoc was able to understand that there would not be a Quarto😅
Anyhow Quarto is "markdown" as far as Pandoc is concerned.

In the case of formatters, you usually want a replacement in place.

**Comment by acederberg** [comment]
> If Pandoc was able to understand that there would not be a Quarto😅
> Anyhow Quarto is "markdown" as far as Pandoc is concerned.

True that. I will just let ``pandoc`` treat ``qmd`` files the same as markdown. I will see what I can cook up and share if it is worth while.

**Reply by cscheid** [reply]
@acederberg These are deep waters, just FYI. You'll need to use a custom Pandoc reader like [we do](https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/pandoc/datadir/readqmd.lua), otherwise Pandoc will not recognize code blocks that use the `{...}` language syntax.
